plutus-chain-index-core/src/Marconi/Index/ScriptTx.hs

```diff
@@ -0,0 +1,194 @@
+    either
+      (const cantDeserialise)
+      (\b -> maybe cantDeserialise (return . ScriptAddress) $ Shelley.deserialiseFromRawBytes Shelley.AsScriptHash b) . deserialiseOrFail
+    where
+      cantDeserialise = SQL.returnError SQL.ConversionFailed f "Cannot deserialise address."
+
+instance SQL.ToRow ScriptTxRow where
+  toRow o = [SQL.toField $ scriptAddress o, SQL.toField $ txCbor o]
+
+-- * Indexer
+
+type Query = ScriptAddress
+type Result = [TxCbor]
+
+data ScriptTxUpdate = ScriptTxUpdate
+  { txScripts :: [(TxCbor, [ScriptAddress])]
+  , slotNo    :: !SlotNo
+  } deriving (Show)
+
+type ScriptTxIndex = SqliteIndex ScriptTxUpdate () Query Result
+
+
+toUpdate :: forall era . C.IsCardanoEra era => [C.Tx era] -> SlotNo -> ScriptTxUpdate
+toUpdate txs = ScriptTxUpdate txScripts'
+  where
+    txScripts' = map (\tx -> (TxCbor $ C.serialiseToCBOR tx, getTxScripts tx)) txs
+
+getTxBodyScripts :: forall era . C.TxBody era -> [ScriptAddress]
+getTxBodyScripts body = let
+    hashesMaybe :: [Maybe C.ScriptHash]
+    hashesMaybe = case body of
+      Shelley.ShelleyTxBody shelleyBasedEra _ scripts _ _ _ -> flip map scripts $ \script ->
+        case fromShelleyBasedScript shelleyBasedEra script of
+          Shelley.ScriptInEra _ script' -> Just $ C.hashScript script'
+      _ -> [] -- Byron transactions have no scripts
+    hashes = catMaybes hashesMaybe :: [Shelley.ScriptHash]
+  in map ScriptAddress hashes
+
+getTxScripts :: forall era . C.Tx era -> [ScriptAddress]
+getTxScripts (C.Tx txBody _ws) = getTxBodyScripts txBody
+
+open :: FilePath -> Depth -> IO ScriptTxIndex
+open dbPath (Depth k) = do
+  ix <- fromJust <$> Ix.newBoxed query store onInsert k ((k + 1) * 2) dbPath
+  let c = ix ^. Ix.handle
+  SQL.execute_ c "CREATE TABLE IF NOT EXISTS script_transactions (scriptAddress TEXT NOT NULL, txCbor BLOB NOT NULL)"
+  SQL.execute_ c "CREATE INDEX IF NOT EXISTS script_address ON script_transactions (scriptAddress)"
+  pure ix
+
+  where
+    onInsert :: ScriptTxIndex -> ScriptTxUpdate -> IO [()]
+    onInsert _ix _update = pure []
+
+    store :: ScriptTxIndex -> IO ()
+    store ix = do
+      persisted <- Ix.getEvents $ ix ^. Ix.storage
+      buffered <- Ix.getBuffer $ ix ^. Ix.storage
+      let updates = buffered ++ persisted :: [ScriptTxUpdate]
+          rows = do
+            ScriptTxUpdate txScriptAddrs _slotNo <- updates
+            (txCbor', scriptAddrs) <- txScriptAddrs
+            scriptAddr <- scriptAddrs
+            pure $ ScriptTxRow scriptAddr txCbor'
+      forM_ rows $
+        SQL.execute (ix ^. Ix.handle) "INSERT INTO script_transactions (scriptAddress, txCbor) VALUES (?, ?)"
+
+    query :: ScriptTxIndex -> Query -> [ScriptTxUpdate] -> IO Result
+    query ix scriptAddress' updates = do
+      persisted :: [SQL.Only TxCbor] <- SQL.query (ix ^. Ix.handle)
+        "SELECT txCbor FROM utxos WHERE scriptAddress = ?" (SQL.Only scriptAddress')
+
+      let
+        buffered :: [TxCbor]
```

Comment:
When querying we also need to account for `buffered` events.

Reply:
Here I get the transaction cbors for a script address from the sqlite database:
https://github.com/input-output-hk/plutus-apps/blob/83ac994f4f7b457f6d5517adca3d590a070fbc09/plutus-chain-index-core/src/Marconi/Index/ScriptTx.hs#L121 and here I get transaction cbors from the buffered events' sequence https://github.com/input-output-hk/plutus-apps/blob/83ac994f4f7b457f6d5517adca3d590a070fbc09/plutus-chain-index-core/src/Marconi/Index/ScriptTx.hs#L127 

I.e I do query both and return them from the function (?).

Reply:
The `updates` parameter only contains the events though. So basically you need to do what you did before for `store`. The way this works right now is that we keep K events in memory plus some buffered events so we can batch updates to the database (which makes inserting rows much more efficient than if we would do it one by one).

So when we store things on disk, we only store the things that are buffered and leave the K blocks in memory (since they can still be rolled back); but when we query we want to account for all events, including the ones that are currently being buffered.

Reply:
Right, fixes here 1886e125f, 800d8f9a8